CollectScenes node

johnhaddon opened the pull request:
This is analogous to the CollectImages node that @danieldresser created recently. It is designed to provide much better performance than a Loop/Parent combo when building scenes dynamically from an arbitrary number of assets.

@donboie, @mattigruener, fancy reviewing since Andrew and Daniel are away? Also, @ivanimanishi if you get a chance to test against the Loop in a real world situation that would be great.

donboie commented on python/GafferSceneTest/CollectScenesTest.py:
```diff
@@ -0,0 +1,167 @@
+		collect["in"].setInput( options["out"] )
+		collect["rootNames"].setValue( IECore.StringVectorData( [ "a", "b" ] ) )
+
+		self.assertEqual( collect["out"]["globals"].getValue()["option:user:test"], IECore.StringData( "a" ) )
```
Why should we have _a_ rather than _b_ in the _option:user:test_ ? Perhaps a comment would be helpful? Ah  the globals come from the first iteration of the collect?

donboie commented on python/GafferSceneUI/CollectScenesUI.py:
```diff
@@ -0,0 +1,86 @@
+
+			Often the rootNames will be driven by an expression that generates
+			a dynamic number of root locations, perhaps by querying an asset
+			management system or listing cache files on disk.
```
Note about the globals from the 0-th iteration?

johnhaddon replied:
I've popped a mention of that into the main node documentation above.

mattigruener commented on src/GafferScene/CollectScenes.cpp:
```diff
@@ -0,0 +1,430 @@
+	{
+		outputs.push_back( outPlug()->getChild<Plug>( input->getName() ) );
+	}
+	else if( input == outPlug()->childNamesPlug() )
```
Out of curiosity - what's the difference between having this block here instead of putting its contents into the block where the rootNamesPlug() is handled? Is that only because we don't want to trigger a recompute if the plug value changes from 'red green blue' to 'red green blue blue'?

johnhaddon replied:
> Out of curiosity - what's the difference between having this block here instead of putting its contents into the block where the rootNamesPlug() is handled? 

The key rule is that the `affects()` method has to match what the `compute()` methods are going to do. In this case, the computes for globals/setNames/set all pull on `outPlug()->childNamesPlug()` rather than on `rootNamesPlug()`, so that is what we need to declare in `affects()`.

In practice, to the outside observer the effects would be pretty similar if we took a shortcut and dirtied directly in the earlier block, and we could probably get away with it. It would just be harder to verify correctness, be misleading to things like DownstreamIterator, and be vulnerable to changes in the implementation of the node if we add further plugs that affect `outPlug()->childNamesPlug()`.

> Is that only because we don't want to trigger a recompute if the plug value changes from 'red green blue' to 'red green blue blue'?

No, the `affects()` method is used to propagate dirtiness through the graph, and at that point we don't consider plug values at all. Dirty propagation is just to notify observers that "hey, something you looked at before has changed". It's then up to the observer to pull any updates it wants by calling `ValuePlug::getValue()`. At that point `hash()` is used to compute the cache key, and then `compute()` is used to compute a fresh result if the cache doesn't contain what we need. So, it's actually `hash()` that is responsible for reusing results. By using the already deduplicated names from `outPlug()->childNames()` in spots like `hashSetNames()` we do get that reuse for the "red green blue blue" case you described, but not because of `affects()`.

mattigruener replied:
Thanks for the explanation! :)

johnhaddon commented on src/GafferScene/CollectScenes.cpp:
```diff
@@ -0,0 +1,430 @@
+				continue;
+			}
+			InternedString childName( *it );
+			if( find( childNames.begin(), childNames.end(), childName ) == childNames.end() )
```
Yeah, good idea. I should benchmark this but I suspect things have to get pretty big before a set becomes a win - InternedString comparisons are really fast (single pointer comparison), and they're all stored contiguously in the vector which has much better locality than a set. I think making a generic algorithm is probably going to be problematic because the threshold at which `set<T>` is a win is so dependent on the size and comparison cost of `T`. Since this affects other nodes (and some of them have slightly different requirements too), can we leave this for a separate PR?

donboie replied:
> Yeah, good idea. I should benchmark this but I suspect things have to get pretty big before a set becomes a win - InternedString comparisons are really fast (single pointer comparison), and they're all stored contiguously in the vector which has much better locality than a set. I think making a generic algorithm is probably going to be problematic because the threshold at which `set<T>` is a win is so dependent on the size and comparison cost of `T`. 

Fine points - feel stoopid for using a set in my example as I just wanted to make sure my point about the template was clear. Couldn't find any pre-existing boost / std mechanism for this.

> Since this affects other nodes (and some of them have slightly different requirements too), can we leave this for a separate PR?

yup

johnhaddon commented on python/GafferSceneTest/__init__.py:
```diff
@@ -126,6 +126,7 @@
 from DeleteFacesTest import DeleteFacesTest
 from DeleteCurvesTest import DeleteCurvesTest
 from DeletePointsTest import DeletePointsTest
+from CollectScenesTest import CollectScenesTest
```
I've rebased it into the correct commit.

johnhaddon commented:
Thanks for the feedback chaps. I've rebased to fix the separation of the ui/tests commits, and added some documentation about the globals to the main node documentation too. Should be good to go...
